bidscoin/dicomsort.py

```diff
@@ -70,27 +71,40 @@ def sortsession(sessionfolder: Path, dicomfiles: list, dicomfield: str, rename:
-            if not patientname:
-                patientname = bids.get_dicomfield('PatientsName', dicomfile)        # This Attribute was/is sometimes called PatientsName?
-
+            # Parse the naming scheme string and test if all attributes are present
+            if re.fullmatch('(({[a-zA-Z]+(:\\d+d)?})|([a-zA-Z0-9_.]+))*', rename_scheme) is None :
+                LOGGER.error('Bad naming scheme. Only alphanumeric characters could be used for the field names (with the optional number of digits afterwards, '
+                             'e.g., \'{InstanceNumber:05d}\'), and only alphanumeric characters, dots, and underscores could be used as separators. ')
+                rename = False
+            else:
+                fields = re.findall('(?<={)([a-zA-Z]+)(?::\\d+d)?(?=})', rename_scheme)
+                field_alternatives = {'InstanceNumber': 'ImageNumber', # alternative field names based on the earlier versions of the Standard or other reasons
+                                      'PatientName': 'PatientsName'}
+                bids_fields = {}
+                for field in fields:
+                    value = bids.get_dicomfield(field, dicomfile)
+                    if not value:
+                        if field in field_alternatives.keys() and bids.get_dicomfield(field_alternatives[field], dicomfile):
+                            rename_scheme.replace('{'+field+'}','{'+field_alternatives[field]+'}')
+                            bids_fields[field_alternatives[field]] = value
+
+                            LOGGER.info('{0} field is absent from the DICOM header, but {1} is present, using {1}'.format(field, field_alternatives[field]))
+                            continue
+
+                        LOGGER.warning(f"Missing '{field}' DICOM field specified in the naming scheme, cannot safely rename {dicomfile}\n")
+                        rename = False
+                    else:
+                        bids_fields[field] = value
+        if not ext:
+            ext = splitext(dicomfile)[1]
         # Move and/or rename the dicomfile in(to) the (series sub)folder
-        if rename and not (patientname and seriesnr and seriesdescr and acquisitionnr and instancenr):
-            LOGGER.warning(f"Missing one or more essential DICOM-fields, cannot safely rename {dicomfile}\n"
-                           f"patientname = {patientname}\n"
-                           f"seriesnumber = {seriesnr}\n"
-                           f"{dicomfield} = {seriesdescr}\n"
-                           f"acquisitionnr = {acquisitionnr}\n"
-                           f"instancenr = {instancenr}")
-            filename = dicomfile.name
-        elif rename:
-            filename = cleanup(f"{patientname}_{seriesnr:03d}_{seriesdescr}_{acquisitionnr:05d}_{instancenr:05d}{ext}")
+        if rename:
+            new_name = rename_scheme.format(**bids_fields)+ext
+            LOGGER.debug('Renaming %s into %s' % (dicomfile.name, new_name))
```

Comment:
Formatted strings rock :-)

Reply:
fixed in 97fdb91123e767014ca25af1602c2f6c3c2a5515